### PR TITLE
feat(18960): Add connectivity and payload checks for behavior and transitions

### PR DESCRIPTION
**Pull request by vanch3d**
See https://hivemq.kanbanize.com/ctrl_board/57/cards/18960/details/

This PR is one part of the "connectivity and payload" epic. 

It implements the UX allowing users to select a `model` in a `Behavior Policy` node and the relevant `event` and `state` in the connected `Transition` node

The design of the feature required a few elements:
- a specification of the Finite State Machine representing the various `models` of a behaviour policy
- two different types of visualisation of the FSM, using `mermaid` and `reactFlow`, aiming at supporting user selection and transition selection
- a custom selector for a single point of access to transitions/states of the selected FSM

Note that the UX is experimental, trying to render what is in effect three interdependent configuration properties (`transition`, `fromState` and `toState`) into a single flow. It doesn't cover all the use cases yet (in particular, the `Any.*` state is not offered)

Also worth noting is that the `reactFlow` visualisation of the FSM, which offers the best potential for user interaction, requires an external layout library (https://github.com/kieler/elkjs) that is almost doubling the size of the runtime bundle!

```
dist/assets/index-b2957c85.js                                1,961.89 kB │ gzip: 623.56 kB
dist/assets/index-5957e81e.js                                3,442.76 kB │ gzip: 1,075.01 kB
```

### Out-of-scope
- `Event.onAny` has been implemented but support for the `Any.*` for states will be addressed in a separate ticket
- FSM visualisation has not been integrated into the flow, since none of them are ready for user interaction. Their future will be dealt with in another ticket

### Before

screenshot-localhost_3000-2024 02 08-09_52_53

### After
screenshot-localhost_3000-2024 02 08-09_46_02

screenshot-localhost_3000-2024 02 08-13_37_16

Screenshot 2024-02-08 at 09 46 52

screenshot-localhost_3000-2024 02 08-09_49_03

### CODE REVIEW

File: hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx [hunk 1]
```diff
@@ -1,29 +1,108 @@
...
+
+  // TODO[NVL] Error messages?
+  const parentPolicy = useMemo(() => {
+    const adapterNode = nodes.find((e) => e.id === selectedNode) as Node<TransitionData> | undefined
+    if (!adapterNode) return undefined
+    const incomers = getIncomers(adapterNode, nodes, edges)
+    if (!incomers) return undefined
+    if (incomers.length !== 1) return undefined
+    const [behavior] = incomers
+    if (behavior.type !== DataHubNodeType.BEHAVIOR_POLICY) return undefined
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx:33]: suggestion: let's update the format a bit as it's very hard to parse as a reader. I suggest the following.


```suggestion
    const adapterNode = nodes.find((node) => node.id === selectedNode) as Node<TransitionData> | undefined
    if (!adapterNode) {
      return undefined
    }
    
    const incomers = getIncomers(adapterNode, nodes, edges)
    if (!incomers || incomers.length !== 1) {
      return undefined
    }

    const [behavior] = incomers
    if (behavior.type !== DataHubNodeType.BEHAVIOR_POLICY) {
      return undefined
    }
```

File: hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx [hunk 2]
```diff
@@ -1,29 +1,108 @@
...
+      const { formData } = initData
+      if (formData) {
+        const { event } = formData
+        if (event) {
+          const [e, from, to] = event.split('-')
+          initData.formData = {
+            ...initData.formData,
+            event: e as unknown as TransitionType,
+            from: from as unknown as StateType,
+            to: to as unknown as StateType,
```
Comment (h2xd) [inline hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx:78]: suggestion: You may can reduce the amount of manual assignments. WDYT about the following

 ```suggestion
           const [event, from, to] = event.split('-') as unknown as [TransitionType, StateType, StateType]
           initData.formData = {
             ...initData.formData,
             event,
             from,
             to,
```

**Review by h2xd — APPROVED** [review]
LGTM! 🐊 

Added couple of comments that focus mainly on readability

File: hivemq-edge/src/frontend/src/extensions/datahub/components/forms/TransitionSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  const onFocus = ({ target: { value } }: FocusEvent<HTMLInputElement>) => props.onFocus(props.id, value)
+
+  const options = useMemo(() => {
+    const metadata = props.options.metadata as FiniteStateMachine | null
+    if (!metadata) return []
+
+    const opts = metadata.transitions.map<FsmTransitionExt>((e) => ({
+      ...e,
+      id: `${e.event}-${e.fromState}-${e.toState}`,
+    }))
```
[the remark this answers is not in the document]
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/forms/TransitionSelect.tsx:85]: I thought i was paying more attention to these :-)

File: hivemq-edge/src/frontend/src/extensions/datahub/components/fsm/FiniteStateMachineFlow.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+export const FiniteStateMachineFlow: FC<FiniteStateMachine> = (props) => {
+  if (import.meta.env.VITE_FLAG_DATAHUB_STATIC_FSM === 'false')
+    return (
+      <Box w="100%" height="400px">
+        <ReactFlowProvider>
+          <ReactFlowRenderer {...props} />
+        </ReactFlowProvider>
+      </Box>
+    )
+  return <MermaidRenderer {...props} />
```
[the remark this answers is not in the document]
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/fsm/FiniteStateMachineFlow.tsx:19]: But I need to cover the case hen it is not set either. I'll reverse the flag, e.g. `VITE_FLAG_DATAHUB_FLOW_FSM`

File: hivemq-edge/src/frontend/src/components/Mermaid.tsx [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  useEffect(() => {
+    mermaid.initialize({ ...DEFAULT_CONFIG })
+  }, [])
+  useEffect(() => {
+    mermaid.contentLoaded()
+  }, [text])
+
+  return (
+    <Card>
+      <CardBody className="mermaid" sx={{ '& svg': { scale: '0.75 0.75' } }} data-processed={undefined}>
```
[the remark this answers is not in the document]
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/components/Mermaid.tsx:59]: Possibly indeed. Mermaid is a Vanilla JS with no good React wrapper, `classname` is the only thing that makes the connection between the React component and the Mermaid renderer. So I deal with the problem by making sure there is only one diagram at a time :-)
And no decision made on usefulness  of that visualisation so it might go at the next PR.

File: hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx (hunk 1 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx:33]: Nice one. ESLint's `curly` rule is not set. Will add it asap.

File: hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx (hunk 2 above)
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx:78]: Do you find that more readable? It's leaner but ambivalent as to whether it reads better. I'll try it, see if it works for me

Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/flow/transition/TransitionPanel.tsx:78]: It works better, thanks to implicit typing and the IDE. And `unknown` is not even needed there

File: hivemq-edge/src/frontend/src/extensions/datahub/components/fsm/FiniteStateMachineFlow.tsx [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { FC } from 'react'
+import { Box } from '@chakra-ui/react'
+
+import { FiniteStateMachine } from '@datahub/types.ts'
+
+import { MermaidRenderer } from '@datahub/components/fsm/MermaidRenderer.tsx'
```
[the remark this answers is not in the document]
Comment (vanch3d) [reply hivemq-edge/src/frontend/src/extensions/datahub/components/fsm/FiniteStateMachineFlow.tsx:6]: Problems with lazy loading. Will be dealt with in [19248](https://hivemq.kanbanize.com/ctrl_board/57/cards/19248/details/)
